kolibri/content/utils/import_export_content.py

```diff
@@ -61,3 +77,40 @@ def get_num_coach_contents(contentnode, filter_available=True):
             .count()
     else:
         return 1 if contentnode.coach_content else 0
+
+
+def retry_import(command, e, name, f=None, overall_progress_update=None):
+    logging.error("An error occured during channel import: {}".format(e))
+
+    # When there is an Internet connection error or timeout error,
+    # or HTTPError where the error code is one of the RETRY_STATUS_CODE,
+    # return False, 0 to retry the file transfer, or return True, 0 to
+    # indicate the cancellation
+    if (
+            isinstance(e, ConnectionError) or
+            isinstance(e, Timeout) or
+            (isinstance(e, HTTPError) and e.response.status_code in RETRY_STATUS_CODE) or
+            (isinstance(e, SSLERROR) and 'decryption failed or bad record mac' in str(e))):
+        return sleep_before_retry(command), 0
+
+    # Skip the file if it does not exist on the server or disk.
+    # This only applies to content import not channel import.
+    elif (
+            name == 'content' and
+            (
+                (isinstance(e, HTTPError) and e.response.status_code == 404) or
+                (isinstance(e, OSError) and e.errno == 2))):
+        overall_progress_update(f.file_size)
+        return True, 1
+
+    else:
+        raise e
+
+def sleep_before_retry(command):
+    logging.info('Waiting for 30 seconds before retrying...')
+    for i in range(30):
+        if command.is_cancelled():
+            command.cancel()
+            return True
+        sleep(1)
```

Comment:
This busy polling is too busy. I think you can sleep for 30 seconds and then check if the job is canceled after that.

The 30 seconds should be a constant expressed somewhere.

Reply:
ah yes, that's true. It makes sense to me. How about checking if the user cancels the import every 10 or 15 seconds instead of doing every second?

Reply:
>  thousands of polls during retries

@benjaoming by 'polls' are you referring to `command.is_cancelled()`? Is this an expensive operation?

Reply:
@indirectlylit likely not, the storage of Iceqube is in memory afaik, but whenever something is busy polling, you are going to see a reaction from someone like me, who doesn't like it :)

For instance, assuming Iceqube is thread-safe, that implies getting a read lock every time you poll something in the job queue.